cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ImpersonationHandler.java

```diff
@@ -0,0 +1,102 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.gateway.handlers;
+
+import co.cask.cdap.common.namespace.NamespacedLocationFactory;
+import co.cask.cdap.data2.security.CredentialsInfo;
+import co.cask.cdap.data2.security.ImpersonationUtils;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.security.DefaultUGIProvider;
+import co.cask.cdap.security.TokenSecureStoreUpdater;
+import co.cask.http.AbstractHttpHandler;
+import co.cask.http.HttpResponder;
+import com.google.inject.Inject;
+import org.apache.hadoop.security.Credentials;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.twill.api.SecureStore;
+import org.apache.twill.filesystem.Location;
+import org.jboss.netty.handler.codec.http.HttpRequest;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.DataOutputStream;
+import java.util.concurrent.Callable;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+/**
+ * Resolves the UGI for a given namespace, acquires the delegation tokens for that UGI,
+ * using {@link TokenSecureStoreUpdater}, and serializes these Credentials to a location.
+ *
+ * Response with the location to which the credentials were serialized to, as well as the UGI's short username
+ */
+// we don't share the same version as other handlers, so we can upgrade/iterate faster
+@Path("/v1/namespaces/{namespace-id}")
+public class ImpersonationHandler extends AbstractHttpHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultUGIProvider.class);
+
+  private final DefaultUGIProvider tokenRequester;
+  private final TokenSecureStoreUpdater tokenSecureStoreUpdater;
+  private final NamespacedLocationFactory namespacedLocationFactory;
+
+  @Inject
+  public ImpersonationHandler(DefaultUGIProvider tokenRequester,
+                              TokenSecureStoreUpdater tokenSecureStoreUpdater,
+                              NamespacedLocationFactory namespacedLocationFactory) {
+    // TODO: only do this stuff in distributed CDAP
+    // TODO: make sure other impersonation stuff only happens in distributed
+    this.tokenRequester = tokenRequester;
+    this.tokenSecureStoreUpdater = tokenSecureStoreUpdater;
+    this.namespacedLocationFactory = namespacedLocationFactory;
+  }
+
+  @POST
+  @Path("/credentials")
+  public void getCredentials(HttpRequest request, HttpResponder responder,
+                             @PathParam("namespace-id") String namespaceId) throws Exception {
```

Comment:
This is called from:
1. DatasetOpExecutor (in order to get delegation tokens for executing DatasetAdmin operations)
2. Will be used from Explore container to get tokens for submitting explore query.
3. Will be used from StreamWriter containers.

This is not called from program containers.